scio-core/src/main/scala/com/spotify/scio/transforms/FilenameRetainingDoFns.scala

```diff
@@ -0,0 +1,165 @@
+    metadata: beam.fs.MatchResult.Metadata,
+    from: Long,
+    to: Long
+  ): beam.FileBasedSource[T] = {
+    // source.createForSubrangeOfFile(file.getMetadata, range.getFrom, range.getTo)
+    SourceSubrangeMethod
+      .invoke(source, metadata, from, to)
+      .asInstanceOf[beam.FileBasedSource[T]]
+  }
+
+  @transient private lazy val SplitIntoRnagesFnCtor = yolo(
+    // TODO harass beam to allow access to this
+    "org.apache.beam.sdk.io.ReadAllViaFileBasedSource$SplitIntoRangesFn",
+    classOf[Long]
+  )
+
+  type SplitIntoRangesT = DoFn[beam.FileIO.ReadableFile, KV[beam.FileIO.ReadableFile, OffsetRange]]
+  def splitIntoRangesFn(desiredBundleSizeBytes: Long): SplitIntoRangesT =
+    SplitIntoRnagesFnCtor.newInstance(desiredBundleSizeBytes).asInstanceOf[SplitIntoRangesT]
+}
+
+object ReadAllViaFileBasedSourceWithFilename {
+  val DefaultReadFileRangesFnExceptionHandler =
+    new beam.ReadAllViaFileBasedSource.ReadFileRangesFnExceptionHandler
+}
+
+/**
+ * Exactly the same as [[beam.ReadAllViaFileBasedSource]] except uses
+ * [[FilenameRetainingReadFileRangesFn]] instead of
+ * [[beam.ReadAllViaFileBasedSource.ReadFileRangesFn]]
+ */
+class ReadAllViaFileBasedSourceWithFilename[T](
+  createSource: SerializableFunction[String, _ <: beam.FileBasedSource[T]],
+  desiredBundleSizeBytes: Long = (64 * 1024 * 1024L), // 64 mb
+  usesReshuffle: Boolean = true,
+  exceptionHandler: beam.ReadAllViaFileBasedSource.ReadFileRangesFnExceptionHandler =
+    ReadAllViaFileBasedSourceWithFilename.DefaultReadFileRangesFnExceptionHandler
+)(implicit
+  coder: Coder[T]
+) extends PTransform[PCollection[ReadableFile], PCollection[(String, T)]] {
+
+  private val outCoder = CoderMaterializer
+    .beamWithDefault(Coder.tuple2Coder[String, T](Coder.stringCoder, coder))
+
+  override def expand(input: PCollection[ReadableFile]): PCollection[(String, T)] = {
+    var ranges: PCollection[KV[ReadableFile, OffsetRange]] = input.apply(
+      "Split into ranges",
+      ParDo.of(AccessibleBeam.splitIntoRangesFn(desiredBundleSizeBytes))
+    )
+    if (usesReshuffle) {
+      ranges = ranges.apply("Reshuffle", Reshuffle.viaRandomKey[KV[ReadableFile, OffsetRange]])
+    }
+    ranges
+      .apply(
+        "Read ranges with filename",
+        ParDo.of(new FilenameRetainingReadFileRangesFn[T](createSource, exceptionHandler))
+      )
+      .setCoder(outCoder)
+  }
+}
+
+/**
+ * Exactly the same as [[beam.ReadAllViaFileBasedSource.ReadFileRangesFn]] except it outputs the
+ * filename in addition to the data contents.
+ */
+class FilenameRetainingReadFileRangesFn[T](
+  createSource: SerializableFunction[String, _ <: beam.FileBasedSource[T]],
+  exceptionHandler: beam.ReadAllViaFileBasedSource.ReadFileRangesFnExceptionHandler
+) extends DoFn[KV[beam.FileIO.ReadableFile, OffsetRange], (String, T)] {
+  type DoFnT = DoFn[KV[beam.FileIO.ReadableFile, OffsetRange], (String, T)]
+
+  @ProcessElement
+  def process(c: DoFnT#ProcessContext): Unit = {
```

Comment:
There is a newer style of implementing process func without ProcessContext, but `@Element`, `OutputReceiver` etc. See https://github.com/spotify/scio/commit/214981e204d3c289f8d5ec751bafa3ad531875e2